Kernel/Tasks/PowerStateSwitchTask.cpp

```diff
@@ -152,15 +155,16 @@ ErrorOr<void> PowerStateSwitchTask::perform_shutdown()
     Processor::halt();
 }
 
-ErrorOr<void> PowerStateSwitchTask::kill_processes(ProcessKind kind, ProcessID finalizer_pid)
+ErrorOr<void> PowerStateSwitchTask::kill_all_user_processes(ProcessID finalizer_pid)
 {
-    bool kill_kernel_processes = kind == ProcessKind::Kernel;
-
-    Process::all_instances().for_each([&](Process& process) {
-        if (process.pid() != Process::current().pid() && process.pid() != finalizer_pid && process.is_kernel_process() == kill_kernel_processes) {
-            process.die();
-        }
-    });
+    {
+        SpinlockLocker lock(g_scheduler_lock);
+        Process::all_instances().for_each([&](Process& process) {
+            if (process.pid() != Process::current().pid() && process.pid() != finalizer_pid && !process.is_kernel_process()) {
```

Comment:
Isn't the finalizer a kernel process?

Reply:
Actually, since we know that this will only ever be called by the power state switch task, can't we just ignore all kernel processes and skip checking the other pids, even our own?

Reply:
Yes, it is actually possible to do so. I will fix this today this evening.